syntax = "proto3";

package paxos;


service Acceptor {
    rpc ReceivePrepare(PrepareRequest) returns (PrepareResponse);
    rpc SendAccept(AcceptRequest) returns (AcceptResponse);
}

service Proposer {
    rpc Prepare(PrepareRequest) returns (PrepareResponse);
    rpc Accept(AcceptRequest) returns (AcceptResponse);
}

message PrepareRequest {
    int64 proposal_number = 1;
}

message PrepareResponse {
    bool promise = 1; 
    int64 proposal_number = 2;
    string accepted_value = 3; 
    int64 accepted_proposal_number = 4; 
}

message AcceptRequest {
    int64 proposal_number = 1;
    string value = 2;  
}

message AcceptResponse {
    bool accepted = 1; 
}

message ProposeRequest {
    string value = 1;
}

message ProposeResponse {
    bool success = 1;
}

message RequestVoteRequest {
    int32 term = 1;
    int32 candidate_id = 2;
}

message RequestVoteResponse {
    int32 term = 1;
    bool vote_granted = 2;
}

message ElectionMessage {
    int32 node_id = 1;
    string message = 2;  // "election", "answer", "coordinator"
}

message ElectionResponse {
    bool ack = 1;
}
